### ISSUE TITLE
pm2 logging problem with multiple applications

### ISSUE THREAD

**Issue by nomiddlename** [nomiddlename/log4js-node#547]
The pm2 logging support breaks when the PM2 instance has multiple applications running within the cluster. Log messages are picked up by all applications instead of the correct ones, resulting in log output going to the wrong files (or not being logged at all).

**Comment by hugeleaf** [comment]
pm2 v2.7.1
node v8.4.0
`log4js.configure({
    appenders: {
        debug: {type: 'console'},
        release: {
            type: 'dateFile',
            filename: config.log_path,
            alwaysIncludePattern: true,
            pattern: '-yyyyMMdd.log',
            daysToKeep: 60,
            layout: {
                type: 'pattern',
                pattern: '[%d{O yyyy-MM-dd hh:mm:ss.SSS}] [%p] %m'
            }
        }
    },
    categories: {
        default: {appenders: ['debug'], level: 'trace'},
        debug: {appenders: ['debug'], level: config.log_level},
        release: {appenders: ['release'], level: config.log_level}
    },
    pm2: true
});`

`┌──────────┬────┬─────────┬───────┬────────┬─────────┬────────┬─────┬───────────┬──────┬──────────┐
│ App name │ id │ mode    │ pid   │ status │ restart │ uptime │ cpu │ mem       │ user │ watching │
├──────────┼────┼─────────┼───────┼────────┼─────────┼────────┼─────┼───────────┼──────┼──────────┤
│ Api1     │ 2  │ cluster │ 31006 │ online │ 2       │ 0s     │ 87% │ 17.2 MB   │ root │ disabled │
│ Api1     │ 3  │ cluster │ 31007 │ online │ 2       │ 0s     │ 89% │ 17.4 MB   │ root │ disabled │
│ Api2     │ 4  │ cluster │ 30977 │ online │ 1       │ 20s    │ 0%  │ 48.2 MB   │ root │ disabled │
│ Api2     │ 5  │ cluster │ 30982 │ online │ 1       │ 20s    │ 1%  │ 47.8 MB   │ root │ disabled │
└──────────┴────┴─────────┴───────┴────────┴─────────┴────────┴─────┴───────────┴──────┴──────────┘
 Module activated
┌───────────────┬─────────┬────────────┬────────┬─────────┬─────┬─────────────┬──────┐
│ Module        │ version │ target PID │ status │ restart │ cpu │ memory      │ user │
├───────────────┼─────────┼────────────┼────────┼─────────┼─────┼─────────────┼──────┤
│ pm2-intercom  │ 1.0.0   │ N/A        │ online │ 0       │ 1%  │ 27.742 MB   │ hitv │
│ pm2-logrotate │ 2.2.0   │ N/A        │ online │ 0       │ 0%  │ 27.918 MB   │ hitv │
└───────────────┴─────────┴────────────┴────────┴─────────┴─────┴─────────────┴──────┘`

issues:
#1 2 apps are launched by PM2 under cluster mode, each app have 2 process. The 2nd app ( id 4,5 )  lose logs, only 1 process log into the file.
#2 2 apps are launched by PM2, 1st app is under fork mode, 2nd is under cluster mode. The 2nd app
lose logs, only 1 process log into the file.
#3 2 apps are launched by PM2, 1st app is under cluster mode, 2nd is under fork mode. Everything is ok.

**Comment by SamriddhaBanik** [comment]
I am also facing the same issue and have found a fix for the same.

Config: 
log4js version 2.5.3
pm2 version 2.9.2
node version 7

when ever we are launching in pm2 cluster mode all my processes are child process but log4j is only attaching to the process with instance ID 0,

if we change the validations, it works fine in cluster mode

changes 
file: log4js.js, line: 152
file: configuration.js, line:83

[configuration_js.txt](https://github.com/log4js-node/log4js-node/files/1889583/configuration_js.txt)
[log4js_js.txt](https://github.com/log4js-node/log4js-node/files/1889584/log4js_js.txt)

this can be triggered based on some flag which allows for existing usage and multi mode usage.

**Comment by nomiddlename** [comment]
I was just following this [document on the PM2 website](http://pm2.keymetrics.io/docs/usage/environment/#specific-environment-variables). I check for a process with instance ID 0 because I only want the appenders to output on one process - you can have problems with multiple processes trying to write to the same files. The other processes should send their log events to the process with ID 0, using pm2-intercom. If you want all processes to behave as master processes, then use the `disableClustering` config option.

**Comment by hugeleaf** [comment]
> The other processes should send their log events to the process with ID 0, using pm2-intercom. 

 ----the  issue is that it does not work when **2 or more APPS** are launched by pm2 under cluster mode,  and pm2-intercom has been installed.

**Comment by SamriddhaBanik** [comment]
Thanks @nomiddlename  adding the param solved the problem.
As for managing multiple apps on the same pm2 instance one thing that can be done is creating a seperate  logging module which will always be listening and all other application would be sending their logs using pm2-intercom. 

but that would lead to pm2-intercom taking up all the resources.

@hugeleaf you can try the solution of making all processes as PM2Masters and check if that helps.
I know its not the correct way, but you can give it a try to check if its working for your configuration.
do not set disableClustering as true and let me know if the same is working.

**Comment by hugeleaf** [comment]
@SamriddhaBanik Thx for your answer, the way you said didn't make me lose logs. But as @nomiddlename said,  there are hidden troubles with multiple processes trying to write to the same files. My current solution is quite simple, output logs from different processes to different files, the log will not be lost anyway🤣

**Comment by nomiddlename** [comment]
@hangaoke1 yes, you are correct. The way to fix this is to namespace the messages from each application, so that messages from the "wrong" applications can be ignored.

**Comment by liujingbreak** [comment]
I just wrote https://github.com/liujingbreak/log4js-pm2-intercom to resolve this problem, but yet not fully tested.

**Comment by sinbargit** [comment]
@nomiddlename Any process or workaround solution? I has this problem too. PM2 cluster mode uncompatible with log4js, that means pm2 can not use at production?

**Comment by newwayxp** [comment]
I have the same problem in 2023, can't believe still not work in latest version. I think maybe add app name when work send message to master, and master check the process name  at first, write to log file only on same name.
